lib/mapObjects/CRewardableObject.cpp

```diff
@@ -826,12 +827,16 @@ void CGOnceVisitable::initObj(CRandomGenerator & rand)
 			onEmpty.addTxt(MetaString::ADVOB_TXT, 38);
 			soundID = soundBase::MYSTERY;
 			blockVisit = true;
+			info.resize(1);
 			if(rand.nextInt(99) < 20)
 			{
-				info.resize(1);
 				loadRandomArtifact(rand, info[0], 10, 10, 10, 0);
 				info[0].message.addTxt(MetaString::ADVOB_TXT, 37);
 			}
+			else
+			{
+				info[0].message.addTxt(MetaString::ADVOB_TXT, 38);
```

Comment:
It is meant to call [grantReward](https://github.com/Chocimier/vcmi/blob/iss1975/lib/mapObjects/CRewardableObject.cpp#L112) which marks object visited.